File: Model Visualizer/Model-Visualizer.c
#include <raylib.h>
#include <stdbool.h>

#define RAYGUI_IMPLEMENTATION
#include "raygui.h"

int main() {
    Vector2 Window = {900, 600};
    SetConfigFlags(FLAG_MSAA_4X_HINT);
    InitWindow(Window.x, Window.y, "Model Visualizer");
    SetTargetFPS(30);

    Model model = LoadModel("Models/SpikeBall.obj");
    Texture2D texture = LoadTexture("Textures/RandomTexture.jpg");
    model.materials[0].maps[0].texture = texture;

    Image backgroundImage = LoadImage("Images/BackgroundImage.jpg");
    ImageResize(&backgroundImage, Window.x, backgroundImage.height * (Window.x / backgroundImage.width));
    Texture2D backgroundTexture = LoadTextureFromImage(backgroundImage);

    Font mainFont = LoadFontEx("Fonts/MainFont.ttf", 100, 0, 250);
    GenTextureMipmaps(&mainFont.texture);
    SetTextureFilter(mainFont.texture, TEXTURE_FILTER_BILINEAR);
    GuiSetFont(mainFont);

    Vector3 modelPos = {0, 0, 0};
    Vector3 modelAxes = {0, 0, 0};
    Vector3 modelScale = {0.5f, 0.5f, 0.5f};

    Camera3D camera = {0};
    camera.position = (Vector3){0, 0, 5};
    camera.target = (Vector3){0, 0, 0};
    camera.up = (Vector3){0, 1, 0};
    camera.fovy = 60;
    camera.projection = CAMERA_PERSPECTIVE;

    GuiSetStyle(DEFAULT, TEXT_COLOR_NORMAL, ColorToInt((Color){255, 255, 255, 240}));
    GuiSetStyle(DEFAULT, TEXT_SIZE, 12);
    GuiSetStyle(DEFAULT, TEXT_SPACING, 0);
    GuiSetStyle(DEFAULT, BASE_COLOR_NORMAL, ColorToInt((Color){255, 255, 255, 10}));
    GuiSetStyle(DEFAULT, BORDER_COLOR_NORMAL, ColorToInt((Color){255, 255, 255, 250}));
    GuiSetStyle(DEFAULT, BASE_COLOR_FOCUSED, ColorToInt((Color){255, 255, 255, 40}));
    GuiSetStyle(DEFAULT, BORDER_COLOR_FOCUSED, ColorToInt((Color){255, 255, 255, 40}));
    GuiSetStyle(DEFAULT, BORDER_WIDTH, 2);
    GuiSetStyle(DEFAULT, BASE_COLOR_PRESSED, ColorToInt((Color){255, 255, 255, 200}));
    GuiSetFont(mainFont);

    float rotationAngle = 0;
    bool toggleWires = true;
    char cameraMode = 0;
    const char *textfile = LoadFileText("float.c");

    while (!WindowShouldClose()) {
        BeginDrawing();
        ClearBackground(BLACK);
        DrawTextureV(backgroundTexture, (Vector2){0, 0}, WHITE);

        const char *infoFPS = TextFormat("%i fps", GetFPS());
        const char *infoScreen = TextFormat("%ix%i Screen", GetScreenWidth(), GetScreenHeight());
        const char *infoRender = TextFormat("%ix%i Render", GetRenderWidth(), GetRenderHeight());
        const char *infoObj = TextFormat("%s", textfile);

        DrawTextEx(mainFont, "Texture loaded successfully", (Vector2){10, 130}, 12, 0, WHITE);
        DrawTextEx(mainFont, model.meshCount == 0 ? "Invalid Model" : "Model loaded successfully",
                   (Vector2){10, 150}, 12, 0, model.meshCount == 0 ? RED : WHITE);
        DrawTextEx(mainFont, backgroundImage.width == 0 || backgroundImage.height == 0 ? "Invalid Image" : "Image loaded successfully",
                   (Vector2){10, 170}, 12, 0, backgroundImage.width == 0 || backgroundImage.height == 0 ? RED : WHITE);
        DrawTextEx(mainFont, mainFont.texture.id == 0 ? "Invalid Font" : "Font loaded successfully",
                   (Vector2){10, 190}, 12, 0, mainFont.texture.id == 0 ? RED : WHITE);

        DrawTextEx(mainFont, infoFPS, (Vector2){10, 10}, 50, 0, WHITE);
        DrawTextEx(mainFont, infoScreen, (Vector2){200, 24}, 14, 0, WHITE);
        DrawTextEx(mainFont, infoRender, (Vector2){200, 44}, 14, 0, WHITE);
        DrawTextEx(mainFont, infoObj, (Vector2){10, 220}, 10, 0, (Color){255, 255, 255, 200});

        switch (cameraMode) {
            case 0: DrawTextEx(mainFont, "Camera Mode: CAMERA_ORBITAL", (Vector2){600, 20}, 12, 0, WHITE); break;
            case 1: DrawTextEx(mainFont, "Camera Mode: CAMERA_FREE", (Vector2){600, 20}, 12, 0, WHITE); break;
            case 2: DrawTextEx(mainFont, "Camera Mode: CAMERA_CUSTOM", (Vector2){600, 20}, 12, 0, WHITE); break;
            case 3: DrawTextEx(mainFont, "Camera Mode: CAMERA_FIRST_PERSON", (Vector2){600, 20}, 12, 0, WHITE); break;
            case 4: DrawTextEx(mainFont, "Camera Mode: CAMERA_THIRD_PERSON", (Vector2){600, 20}, 12, 0, WHITE); break;
        }

        BeginMode3D(camera);
        if (toggleWires)
            DrawModelEx(model, modelPos, modelAxes, rotationAngle, modelScale, WHITE);
        else
            DrawModelWiresEx(model, modelPos, modelAxes, rotationAngle, modelScale, WHITE);
        EndMode3D();

        if (IsKeyDown(KEY_W) || IsKeyDown(KEY_Z)) GuiButton((Rectangle){10, 80, 50, 30}, GuiIconText(ICON_ARROW_UP, "W/Z"));
        if (IsKeyDown(KEY_S)) GuiButton((Rectangle){10, 80, 50, 30}, GuiIconText(ICON_ARROW_DOWN, "S"));
        if (IsKeyDown(KEY_D)) GuiButton((Rectangle){10, 80, 50, 30}, GuiIconText(ICON_ARROW_RIGHT, "D"));
        if (IsKeyDown(KEY_A) || IsKeyDown(KEY_Q)) GuiButton((Rectangle){10, 80, 50, 30}, GuiIconText(ICON_ARROW_LEFT, "Q/A"));
        if (IsKeyDown(KEY_UP)) GuiButton((Rectangle){70, 80, 100, 30}, GuiIconText(ICON_ARROW_UP, "UP ARROW"));
        if (IsKeyDown(KEY_DOWN)) GuiButton((Rectangle){70, 80, 100, 30}, GuiIconText(ICON_ARROW_DOWN, "DOWN ARROW"));
        if (IsKeyDown(KEY_RIGHT)) GuiButton((Rectangle){70, 80, 100, 30}, GuiIconText(ICON_ARROW_RIGHT, "RIGHT ARROW"));
        if (IsKeyDown(KEY_LEFT)) GuiButton((Rectangle){70, 80, 100, 30}, GuiIconText(ICON_ARROW_LEFT, "LEFT ARROW"));

        if (GuiButton((Rectangle){600, 500, 120, 30}, GuiIconText(ICON_REREDO_FILL, "Reset"))) {
            modelPos = (Vector3){0, 0, 0};
            modelAxes = (Vector3){0, 0, 0};
            modelScale = (Vector3){0.5f, 0.5f, 0.5f};
            camera.position = (Vector3){0, 0, 5};
            camera.target = (Vector3){0, 0, 0};
            camera.up = (Vector3){0, 1, 0};
            rotationAngle = 0;
            camera.fovy = 60;
        }

        if (GuiButton((Rectangle){730, 500, 120, 30}, GuiIconText(ICON_REPEAT, "Show Wires")))
            toggleWires = !toggleWires;

            if (GuiButton((Rectangle){600, 550, 250, 30}, GuiIconText(ICON_CAMERA, "Camera")))
                cameraMode = (cameraMode + 1) % 5;

                GuiSliderBar((Rectangle){600, 100, 200, 10}, "Position X", TextFormat("%.2ff", modelPos.x), &modelPos.x, -10, 10);
                GuiSliderBar((Rectangle){600, 115, 200, 10}, "Position Y", TextFormat("%.2ff", modelPos.y), &modelPos.y, -10, 10);
                GuiSliderBar((Rectangle){600, 130, 200, 10}, "Position Z", TextFormat("%.2ff", modelPos.z), &modelPos.z, -10, 10);
                GuiSliderBar((Rectangle){600, 155, 200, 10}, "Rotation Axe X", TextFormat("%.2ff", modelAxes.x), &modelAxes.x, -10, 10);
                GuiSliderBar((Rectangle){600, 170, 200, 10}, "Rotation Axe Y", TextFormat("%.2ff", modelAxes.y), &modelAxes.y, -10, 10);
                GuiSliderBar((Rectangle){600, 185, 200, 10}, "Rotation Axe Z", TextFormat("%.2ff", modelAxes.z), &modelAxes.z, -10, 10);
                GuiSliderBar((Rectangle){600, 210, 200, 10}, "Scale X", TextFormat("%.2ff", modelScale.x), &modelScale.x, -10, 10);
                GuiSliderBar((Rectangle){600, 225, 200, 10}, "Scale Y", TextFormat("%.2ff", modelScale.y), &modelScale.y, -10, 10);
                GuiSliderBar((Rectangle){600, 240, 200, 10}, "Scale Z", TextFormat("%.2ff", modelScale.z), &modelScale.z, -10, 10);
                GuiSliderBar((Rectangle){600, 265, 200, 10}, "Camera Position X", TextFormat("%.2ff", camera.position.x), &camera.position.x, -10, 10);
                GuiSliderBar((Rectangle){600, 280, 200, 10}, "Camera Position Y", TextFormat("%.2ff", camera.position.y), &camera.position.y, -10, 10);
                GuiSliderBar((Rectangle){600, 295, 200, 10}, "Camera Position Z", TextFormat("%.2ff", camera.position.z), &camera.position.z, -10, 10);
                GuiSliderBar((Rectangle){600, 310, 200, 10}, "Camera Target X", TextFormat("%.2ff", camera.target.x), &camera.target.x, -10, 10);
                GuiSliderBar((Rectangle){600, 325, 200, 10}, "Camera Target Y", TextFormat("%.2ff", camera.target.y), &camera.target.y, -10, 10);
                GuiSliderBar((Rectangle){600, 340, 200, 10}, "Camera Target Z", TextFormat("%.2ff", camera.target.z), &camera.target.z, -10, 10);
                GuiSliderBar((Rectangle){600, 365, 200, 10}, "Camera Up X", TextFormat("%.2ff", camera.up.x), &camera.up.x, -10, 10);
                GuiSliderBar((Rectangle){600, 380, 200, 10}, "Camera Up Y", TextFormat("%.2ff", camera.up.y), &camera.up.y, -10, 10);
                GuiSliderBar((Rectangle){600, 395, 200, 10}, "Camera Up Z", TextFormat("%.2ff", camera.up.z), &camera.up.z, -10, 10);
                GuiSliderBar((Rectangle){600, 420, 200, 10}, "FOVY", TextFormat("%.2ff", camera.fovy), &camera.fovy, -100, 100);
                GuiSliderBar((Rectangle){600, 445, 200, 10}, "Rotation Angle", TextFormat("%.2ff", rotationAngle), &rotationAngle, 0, 360);

                EndDrawing();

                switch (cameraMode) {
                    case 0: UpdateCamera(&camera, CAMERA_ORBITAL); break;
                    case 1: UpdateCamera(&camera, CAMERA_FREE); break;
                    case 2: UpdateCamera(&camera, CAMERA_CUSTOM); break;
                    case 3: UpdateCamera(&camera, CAMERA_FIRST_PERSON); break;
                    case 4: UpdateCamera(&camera, CAMERA_THIRD_PERSON); break;
                }
    }

    UnloadTexture(backgroundTexture);
    UnloadModel(model);
    UnloadImage(backgroundImage);
    UnloadFont(mainFont);
    CloseWindow();
    return 0;
}

File: Model Visualizer/Model-Visualizer.py
import pyray as rl

def main():
    window = rl.Vector2(900, 600)
    rl.set_config_flags(rl.FLAG_MSAA_4X_HINT)
    rl.init_window(int(window.x), int(window.y), b"Model Visualizer")
    rl.set_target_fps(30)

    model = rl.load_model(b"Models/SpikeBall.obj")
    texture = rl.load_texture(b"Textures/RandomTexture.jpg")
    model.materials[0].maps[0].texture = texture

    background_image = rl.load_image(b"Images/BackgroundImage.jpg")
    scale_factor = window.x / background_image.width
    rl.image_resize(background_image, int(window.x), int(background_image.height * scale_factor))
    background_texture = rl.load_texture_from_image(background_image)

    main_font = rl.load_font_ex(b"Fonts/MainFont.ttf", 100, None, 250)
    rl.gen_texture_mipmaps(main_font.texture)
    rl.set_texture_filter(main_font.texture, rl.TEXTURE_FILTER_BILINEAR)
    rl.gui_set_font(main_font)

    model_pos = rl.Vector3(0, 0, 0)
    model_axes = rl.Vector3(0, 0, 0)
    model_scale = rl.Vector3(0.5, 0.5, 0.5)

    camera = rl.Camera3D()
    camera.position = rl.Vector3(0, 0, 5)
    camera.target = rl.Vector3(0, 0, 0)
    camera.up = rl.Vector3(0, 1, 0)
    camera.fovy = 60
    camera.projection = rl.CAMERA_PERSPECTIVE

    rl.gui_set_style(rl.DEFAULT, rl.TEXT_COLOR_NORMAL, rl.color_to_int(rl.Color(255, 255, 255, 240)))
    rl.gui_set_style(rl.DEFAULT, rl.TEXT_SIZE, 12)
    rl.gui_set_style(rl.DEFAULT, rl.TEXT_SPACING, 0)
    rl.gui_set_style(rl.DEFAULT, rl.BASE_COLOR_NORMAL, rl.color_to_int(rl.Color(255, 255, 255, 10)))
    rl.gui_set_style(rl.DEFAULT, rl.BORDER_COLOR_NORMAL, rl.color_to_int(rl.Color(255, 255, 255, 250)))
    rl.gui_set_style(rl.DEFAULT, rl.BASE_COLOR_FOCUSED, rl.color_to_int(rl.Color(255, 255, 255, 40)))
    rl.gui_set_style(rl.DEFAULT, rl.BORDER_COLOR_FOCUSED, rl.color_to_int(rl.Color(255, 255, 255, 40)))
    rl.gui_set_style(rl.DEFAULT, rl.BORDER_WIDTH, 2)
    rl.gui_set_style(rl.DEFAULT, rl.BASE_COLOR_PRESSED, rl.color_to_int(rl.Color(255, 255, 255, 200)))
    rl.gui_set_font(main_font)

    rotation_angle = 0.0
    toggle_wires = True
    camera_mode = 0
    textfile = rl.load_file_text(b"float.c")

    while not rl.window_should_close():
        rl.begin_drawing()
        rl.clear_background(rl.BLACK)
        rl.draw_texture_v(background_texture, rl.Vector2(0, 0), rl.WHITE)

        info_fps = f"{rl.get_fps()} fps"
        info_screen = f"{rl.get_screen_width()}x{rl.get_screen_height()} Screen"
        info_render = f"{rl.get_render_width()}x{rl.get_render_height()} Render"
        info_obj = textfile.decode() if textfile else ""

        rl.draw_text_ex(main_font, b"Texture loaded successfully", rl.Vector2(10, 130), 12, 0, rl.WHITE)
        rl.draw_text_ex(main_font,
                        b"Invalid Model" if model.mesh_count == 0 else b"Model loaded successfully",
                        rl.Vector2(10, 150), 12, 0, rl.RED if model.mesh_count == 0 else rl.WHITE)
        rl.draw_text_ex(main_font,
                        b"Invalid Image" if background_image.width == 0 or background_image.height == 0 else b"Image loaded successfully",
                        rl.Vector2(10, 170), 12, 0, rl.RED if background_image.width == 0 or background_image.height == 0 else rl.WHITE)
        rl.draw_text_ex(main_font,
                        b"Invalid Font" if main_font.texture.id == 0 else b"Font loaded successfully",
                        rl.Vector2(10, 190), 12, 0, rl.RED if main_font.texture.id == 0 else rl.WHITE)

        rl.draw_text_ex(main_font, info_fps.encode(), rl.Vector2(10, 10), 50, 0, rl.WHITE)
        rl.draw_text_ex(main_font, info_screen.encode(), rl.Vector2(200, 24), 14, 0, rl.WHITE)
        rl.draw_text_ex(main_font, info_render.encode(), rl.Vector2(200, 44), 14, 0, rl.WHITE)
        rl.draw_text_ex(main_font, info_obj.encode(), rl.Vector2(10, 220), 10, 0, rl.Color(255, 255, 255, 200))

        modes = [
            b"Camera Mode: CAMERA_ORBITAL",
            b"Camera Mode: CAMERA_FREE",
            b"Camera Mode: CAMERA_CUSTOM",
            b"Camera Mode: CAMERA_FIRST_PERSON",
            b"Camera Mode: CAMERA_THIRD_PERSON"
        ]
        rl.draw_text_ex(main_font, modes[camera_mode], rl.Vector2(600, 20), 12, 0, rl.WHITE)

        rl.begin_mode3d(camera)
        if toggle_wires:
            rl.draw_model_ex(model, model_pos, model_axes, rotation_angle, model_scale, rl.WHITE)
        else:
            rl.draw_model_wires_ex(model, model_pos, model_axes, rotation_angle, model_scale, rl.WHITE)
        rl.end_mode3d()

        # Camera controls
        if rl.is_key_down(rl.KEY_W) or rl.is_key_down(rl.KEY_Z):
            rl.gui_button(rl.Rectangle(10, 80, 50, 30), b"W/Z")
        if rl.is_key_down(rl.KEY_S):
            rl.gui_button(rl.Rectangle(10, 80, 50, 30), b"S")
        if rl.is_key_down(rl.KEY_D):
            rl.gui_button(rl.Rectangle(10, 80, 50, 30), b"D")
        if rl.is_key_down(rl.KEY_A) or rl.is_key_down(rl.KEY_Q):
            rl.gui_button(rl.Rectangle(10, 80, 50, 30), b"Q/A")

        if rl.gui_button(rl.Rectangle(600, 500, 120, 30), b"Reset"):
            model_pos = rl.Vector3(0, 0, 0)
            model_axes = rl.Vector3(0, 0, 0)
            model_scale = rl.Vector3(0.5, 0.5, 0.5)
            camera.position = rl.Vector3(0, 0, 5)
            camera.target = rl.Vector3(0, 0, 0)
            camera.up = rl.Vector3(0, 1, 0)
            rotation_angle = 0
            camera.fovy = 60

        if rl.gui_button(rl.Rectangle(730, 500, 120, 30), b"Show Wires"):
            toggle_wires = not toggle_wires

        if rl.gui_button(rl.Rectangle(600, 550, 250, 30), b"Camera"):
            camera_mode = (camera_mode + 1) % 5

        rl.end_drawing()

        if camera_mode == 0:
            rl.update_camera(camera, rl.CAMERA_ORBITAL)
        elif camera_mode == 1:
            rl.update_camera(camera, rl.CAMERA_FREE)
        elif camera_mode == 2:
            rl.update_camera(camera, rl.CAMERA_CUSTOM)
        elif camera_mode == 3:
            rl.update_camera(camera, rl.CAMERA_FIRST_PERSON)
        elif camera_mode == 4:
            rl.update_camera(camera, rl.CAMERA_THIRD_PERSON)

    rl.unload_texture(background_texture)
    rl.unload_model(model)
    rl.unload_image(background_image)
    rl.unload_font(main_font)
    rl.close_window()

if __name__ == "__main__":
    main()
